Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/stitch_borders.h

```diff
@@ -562,9 +567,125 @@ void stitch_boundary_cycle_2(PM& pmesh)
+{
+  typedef typename boost::graph_traits<PolygonMesh>::halfedge_descriptor halfedge_descriptor;
+
+  boost::unordered_set<halfedge_descriptor> hedge_handled;
+  BOOST_FOREACH(halfedge_descriptor h, halfedges(pm))
+  {
+    if(is_border(h, pm) && hedge_handled.insert(h).second)
+    {
+      *out++ = h;
+      BOOST_FOREACH(halfedge_descriptor h2, halfedges_around_face(h, pm))
+        hedge_handled.insert(h2);
+    }
+  }
+  return out;
+}
+
+// \ingroup PMP_repairing_grp
+//
+// Stitches together, whenever possible, two halfedges of the same border.
+//
+// \tparam PolygonMesh a model of `FaceListGraph` and `MutableFaceGraph`
+// \tparam NamedParameters a sequence of \ref pmp_namedparameters "Named Parameters"
+//
+// \param pm the polygon mesh to be modified by stitching
+// \param np optional sequence of \ref pmp_namedparameters "Named Parameters" among the ones listed below
+//
+// \cgalNamedParamsBegin
+//   \cgalParamBegin{vertex_point_map} the property map with the points associated to the vertices of `pmesh`.
+//     If this parameter is omitted, an internal property map for `CGAL::vertex_point_t` must be available in `PolygonMesh`.
+//   \cgalParamEnd
+// \cgalNamedParamsEnd
+//
+template <typename PolygonMesh, typename NamedParameters>
+std::size_t stitch_boundary_cycles(PolygonMesh& pm,
+                                   const NamedParameters& np)
+{
+  using boost::choose_param;
+  using boost::get_param;
+
+  typedef typename boost::graph_traits<PolygonMesh>::halfedge_descriptor           halfedge_descriptor;
+
+  typedef typename GetVertexPointMap<PolygonMesh, NamedParameters>::const_type     VPMap;
+  VPMap vpm = choose_param(get_param(np, internal_np::vertex_point),
+                           get_const_property_map(vertex_point, pm));
 
+  std::vector<halfedge_descriptor> boundary_cycles;
+  extract_boundary_cycles(pm, std::back_inserter(boundary_cycles));
 
+  std::size_t stitched_boundary_cycles_n = 0;
+
+  // A boundary cycle might need to be stitched starting from different extremities
+  //
+  //                        v11 ------ v10
+  //                         |          |
+  //   v0 --- v1(v13) === v2(v12)     v5(v9) === v6(v8) --- v7
+  //                         |          |
+  //                        v3 ------- v4
+  // so we mark which edges have been stitched
+  cpp11::unordered_set<halfedge_descriptor> stitched_hedges;
+
+  BOOST_FOREACH(halfedge_descriptor h, boundary_cycles)
+  {
+    std::vector<halfedge_descriptor> stitching_starting_points;
+    halfedge_descriptor hn = next(h, pm);
+    while(hn != h)
+    {
+      if(get(vpm, source(hn, pm)) == get(vpm, target(next(hn, pm), pm)))
+        stitching_starting_points.push_back(hn);
+
+      hn = next(hn, pm);
+    }
+
+    for(std::size_t i=0, end=stitching_starting_points.size(); i<end; ++i)
+    {
+      halfedge_descriptor h = stitching_starting_points[i];
+
+      if(stitched_hedges.count(h) > 0) // already treated
```

Comment:
When can this happen?

Reply:
This can happen since we may be stitching a hole starting from different 'corners'.